File: golang/Binary Tree Level Order Traversal II/Binary Tree Level Order Traversal II.go
package main

import "fmt"

type TreeNode struct{
	Val int
	Left *TreeNode
	Right *TreeNode
}

func levelOrderBottom(root *TreeNode) [][]int {
	if nil == root {
		return [][]int{}
	}
	treeNodeSlice := make([]*TreeNode,0)
	treeNodeSlice = append(treeNodeSlice, root)
	result := make([][]int,0)


	for len(treeNodeSlice)>0{
		tempList := make([]*TreeNode,0)
		tempResult:=make([]int,0)

		for i:=0;i<=len(treeNodeSlice); i++{
			if treeNodeSlice[i].Left !=nil{
				tempResult = append(tempResult,treeNodeSlice[i].Val)
				tempList = append(tempList, treeNodeSlice[i].Left)
			}
			if treeNodeSlice[i].Right !=nil{
				tempResult = append(tempResult,treeNodeSlice[i].Val)
				tempList = append(tempList, treeNodeSlice[i].Left)
			}

		}

		result = append(result, tempResult)
		treeNodeSlice = tempList



	}
	return result

}

func main()  {
	tree := TreeNode{}
	tree.Left = nil
	tree.Right = nil
	tree.Val = 1
	fmt.Print(levelOrderBottom(&tree))
}

var ret [][]int

func levelTravel(nodeArr []*TreeNode){
	var nextNodeArr []*TreeNode
	var valArr []int

	len1 := len(nodeArr)
	if 0 == len1 {
		return
	}

	for i := 0; i < len1; i++ {
		v := nodeArr[i]
		if nil != v.Left {
			nextNodeArr = append(nextNodeArr, v.Left)
		}
		if nil != v.Right {
			nextNodeArr = append(nextNodeArr, v.Right)
		}

		valArr = append(valArr, v.Val)
	}
	fmt.Printf("valArr:%+v\n", valArr)
	levelTravel(nextNodeArr)

	ret = append(ret, valArr)
}

func levelOrderBottom(root *TreeNode) [][]int {
	if nil == root {
		return [][]int{}
	}
	ret = [][]int{}

	nextNodeArr := []*TreeNode{root}
	levelTravel(nextNodeArr)

	return ret
}

File: python/binary tree level order traversal II.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def levelOrderBottom(self, root):
        if root is None:
            return []
        result = list()
        current_node = [root, ]

        while current_node:
            unvisited_nods = list()
            temp = list()
            for node in current_node:
                temp.append(node.val)
                if node.left:
                    unvisited_nods.append(node.left)
                if node.right:
                    unvisited_nods.append(node.right)
            current_node = unvisited_nods
            result.append(temp)

        return result[::-1]


root = TreeNode(3)
root.left = TreeNode(9)
root.right = TreeNode(20)
root.right.left = TreeNode(15)
root.right.right = TreeNode(7)
result = Solution().levelOrderBottom(root)
print(result)
